### DISCUSSION TITLE
Limit clients that can access this node

### DISCUSSION THREAD

**Discussion by bartbutenaers** [mikejac/node-red-contrib-google-smarthome#596] [Ideas]
Hi everybody,

I am looking for a way to limit the number of clients that can access this service, as an extra layer of security.
The most optimal scenario would be that only the Google Assistant servers would get access to it.
However that won't be possible, since their ip address will change continiously.

A colleague of mine proposed to to use the Google [ASN](https://www.thousandeyes.com/learning/glossary/as-autonomous-system) number (AS15169), which you can use periodically to get a list of all their IP address ranges.  However when you look at that ASN number, it [seems](https://mxtoolbox.com/SuperTool.aspx?action=asn%3aAS15169&run=toolpage) that they have 12K IP addresses available in their ranges.  Which means that whitelisting their entire ASN, allows Google’s APIs to access this node but also allows all Google Cloud Platform customers to access it. So that wouldn't be very useful as an extra layer of protection.

However I found an more granular way of working in [this](https://coreybraun.com/posts/google-assistant-home-assistant/) tutorial (which is base on [this](https://cloud.google.com/vpc/docs/configure-private-google-access#ip-addr-defaults) python script):
1. They get a list of all Google ip address ranges, based on the ASN number.
2. They get a list of all ip address ranges for the Google Cloud platform customers.
3. They substract list 2 from list 1, and as a result they have a list of all ip address ranges for the Google API servers.

I know that security optimal should be put _**in front**_ of a local service (e.g. using a reverse proxy) instead of _**inside**_ the local service.  But the above logic seems to be very specific for Google.  Therefore I would be tempted to put this kind of logic (as js code obviously) inside this node, as an optional feature that can be turned on/off.

So I would like to hear some thoughts about this!!

P.S. obviously I would contribute to the development of such a feature.  Although I wouldn't manage in my limited free time without the help from people that are familiar with this codebase.

Thanks for reading,
Bart

**Comment by bartbutenaers** [comment]
Remark: while I was developing a proof of concept last night, I got stuck with [this](https://github.com/silverwind/cidr-tools/issues/24) issue.  Hopefully the author of that library can find some time to assist me with that, because I need it to continue here.

**Comment by Caprico85** [comment]
Hi,

I admit I'm not entirely happy with this. I totally agree that it improves security. And it is perfectly fine if you know what you are doing. Unfortunately, not everyone knows what they are doing. People think, "Letting only Google access my service sounds good. Who else should access my service? What can go wrong?" But allowing only Google also means you are blocking yourself. You can't access your service while you're debugging, and you can't log in again without unblocking first. This can lead to more frustrated users and more Github issues.

However, since we have already had several requests for this functionality, I would still consider including it if the user gets some reminders that blocking is enabled.


So I think we would need:
- Middleware for the Express web server to block requests from unknown IP addresses
- A checkbox in the configuration to enable or disable blocking
- Maybe a note at the top of the configuration if filtering is enabled
- Get the current list of Google addresses at startup
- Set an interval to regularly update the list
- Expand the documentation:
  - Add a note in the troubleshooting section to disable blocking if you experience problems
  - Explain the field in google-smarthome.html where all config options are explained
  - In the setup guide, explain the field, but ask to disable it first and only enable it later if everything works.


And some other thoughts:

- Perhaps a first step would be to filter on all Google IPs, without subtracting the Google Cloud Platform IPs. Of course, this would not filter out all unwanted traffic, but at least some of it. And we would see if this is a viable approach.
- How should we block the request? I'm thinking of responding with an HTTP 403 and a message that the request has been blocked because Google filtering is enabled. If users see this message in the browser or in the logs, they may know what they are doing wrong.
- How often do we have to update the list? Is daily enough? Is it possible for Google to add a new address to the list and start using it minutes later?
- Should we exclude the login page and review page from blocking? These pages are accessed directly by the user, not by a Google server. There is a security risk in excluding these pages from blocking, as they can be used to brute-force username/password. Including these pages in the block means that users will have to temporarily disable the block when logging in. 
- Should we exclude internal URLs (192.168.0.0, 10.0.0.0, ...) from blocking? They are needed for local fulfilment and debugging.
- Should the service get it's own IP address from somewhere and whitelist it? This might help at least some of the users who are blocking themselves.

**Comment by bartbutenaers** [comment]
@Caprico85,

Thanks for the extensive feedback!

+ Yes you are right that users should be able to whitelist some other ip ranges, like e.g. the home network `192.168.1.0/24`.
+ I agree with the bullets about your *"So I think we would need:"* list.
+ We could indeed in a first phase allow all Google IP addresses, but that means if someone rents a box in the Google cloud that he gets acesss to your port.  So then we truly give a false sense of security to users imho.
+ Yes your message about blocking because *"Google filtering is enabled"* is something we could easily explain and provide a solution for in your troubleshooting tips.
+ It indeed sounds to make sense to block also the login page and review page.  Why do you think the users have to disable the block temporarily?  I am not familiar with that level of details about your node unfortunately.
+ Should we exclude internal URLs (192.168.0.0, 10.0.0.0, ...) from blocking.  Sounds very logical.  But perhaps instead of hardcoding those ranges, we can perhaps show them on your config screen (e.g. as an array in a typedinput of type json)?
+ Could you please explain *"Should the service get it's own IP address from somewhere and whitelist it?"* a bit more?  Not sure if I understand it correctly.

> Remark: while I was developing a proof of concept last night, I got stuck with https://github.com/silverwind/cidr-tools/issues/24 issue. Hopefully the author of that library can find some time to assist me with that, because I need it to continue here.

The author of the cidr-tools library has today confirmed that it is a bug.  So hopefully it will be fixed soon, so I can continue experimenting.

**Comment by jmarcgit** [comment]
Hi,

You maybe interested in that : [https://github.com/jmarcgit/google-nftables](https://github.com/jmarcgit/google-nftables)

It generates a Google servers white list for nftables.  It's a nodejs script that can be scheduled with cron. I'm using a library called fast-cidr-tools for the range calculations. This library must be installed with npm globally. There is a bug with ipv6 addresses but I'm not even sure that ipv6 is supported right now so it's disabled for now.
